xapian-letor/api/feature_internal.cc

```diff
@@ -0,0 +1,181 @@
+/** @file feature_internal.cc
+ * @brief Definition of Feature::Internal class.
+ */
+/* Copyright (C) 2019 Vaibhav Kansagara
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian-letor/feature.h"
+#include "feature_internal.h"
+
+#include <cmath>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include "debuglog.h"
+
+using namespace std;
+using namespace Xapian;
+
+Database
+Feature::Internal::get_database() const
+{
+    LOGCALL(API, Database, "Feature::Internal::get_database", NO_ARGS);
+    return feature_db;
+}
+
+Query
+Feature::Internal::get_query() const
+{
+    LOGCALL(API, Query, "Feature::Internal::get_query", NO_ARGS);
+    return feature_query;
+}
+
+Document
+Feature::Internal::get_doc() const
+{
+    LOGCALL(API, Document, "Feature::Internal::get_doc", NO_ARGS);
+    return feature_doc;
+}
+
+bool
+Feature::Internal::get_termfreq(const std::string & term,
+			double & termfreq_) const
+{
+    LOGCALL(API, bool, "Feature::Internal::get_termfreq", term | termfreq_);
+    auto termfreq_iterator = termfreq.find(term);
+    if (termfreq_iterator != termfreq.end()) {
+	termfreq_ = (double)termfreq_iterator->second;
+	return true;
+    }
+    return false;
+}
+
+bool
+Feature::Internal::get_inverse_doc_freq(const std::string & term,
+			double & inverse_doc_freq_) const
+{
+    LOGCALL(API, bool, "Feature::Internal::get_inverse_doc_freq", term | inverse_doc_freq_);
+    auto inverse_doc_freq_iterator = inverse_doc_freq.find(term);
+    if (inverse_doc_freq_iterator != inverse_doc_freq.end()) {
+	inverse_doc_freq_ = (double)inverse_doc_freq_iterator->second;
```

Comment:
This cast shouldn't be needed, as the type of the values in the map should be `double`.

If a cast were needed, please avoid C-style casts - instead use a C++-style cast which explains the intent of the cast so that the compiler can reject conversions other than the intended one (here that would be `static_cast<double>(...)`), or for a case like this a C++ "constructor cast" like `double(...)` is OK.

Reply:
fixed in a fixup commit.